ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -574,6 +588,182 @@ releaseResources rr keys releaser = do
+        , tempRegistryResource = remainingResource
+        }
+    return a
+
+-- | When 'runWithTempRegistry' exits successfully while there are still
+-- resources remaining in the temporary registry that haven't been transferred
+-- to the final state.
+data TempRegistryException =
+    forall m. IOLike m => TempRegistryRemainingResource {
+        -- | The context in which the temporary registry was created.
+        tempRegistryContext  :: !(Context m)
+
+        -- | The context in which the resource was allocated that was not
+        -- transferred to the final state.
+      , tempRegistryResource :: !(Context m)
+      }
+
+deriving instance Show TempRegistryException
+instance Exception TempRegistryException
+
+-- | Given a final state, return the 'ResourceId's of the resources that have
+-- been /transferred to/ that state.
+newtype TransferredTo st = TransferredTo {
+      runTransferredTo :: st -> Set ResourceId
+    }
+  deriving newtype (Semigroup, Monoid)
+  deriving NoUnexpectedThunks via OnlyCheckIsWHNF "TransferredTo" (TransferredTo st)
+
+-- | The environment used to run a 'WithTempRegistry' action.
+data TempRegistry st m = TempRegistry {
+      tempResourceRegistry :: !(ResourceRegistry m)
+    , tempTransferredTo    :: !(StrictTVar m (TransferredTo st))
+      -- ^ Used as a @Writer@.
+    }
+
+-- | An action with a temporary registry in scope, see 'runWithTempRegistry'
+-- for more details.
+--
+-- The most important function to run in this monad is 'allocateTemp'.
+newtype WithTempRegistry st m a = WithTempRegistry {
+      unWithTempRegistry :: ReaderT (TempRegistry st m) m a
+    }
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch, MonadMask, MonadSTM)
+
+instance MonadTrans (WithTempRegistry st) where
+  lift = WithTempRegistry . lift
+
+instance MonadState s m => MonadState s (WithTempRegistry st m) where
+  state = WithTempRegistry . state
+
+-- | Untrack all the resources from the registry that have been transferred to
+-- the given state.
+--
+-- Untracking a resource means removing it from the registry without releasing
+-- it.
+--
+-- NOTE: does not check that it's called by the same thread that allocated the
+-- resources, as it's an internal function only used in 'runWithTempRegistry'.
+untrackTransferredTo
+  :: IOLike m
+  => ResourceRegistry m
+  -> TransferredTo st
+  -> st
+  -> m ()
+untrackTransferredTo rr transferredTo st =
+    updateState rr $ mapM_ removeResource rids
+  where
+    rids = runTransferredTo transferredTo st
+
+-- | Allocate a resource in a temporary registry until it has been transferred
+-- to the final state @st@. See 'runWithTempRegistry' for more details.
+allocateTemp
+  :: (IOLike m, HasCallStack)
+  => m a
+     -- ^ Allocate the resource
+  -> (a -> m Bool)
+     -- ^ Release the resource, return 'True' when the resource was actually
+     -- released, return 'False' when the resource was already released.
```

Comment:
Mention again (or refer to `allocateEither`) that the user is allowed to always return `True`.